src/fidesops/service/authentication/authentication_strategy_oauth2.py

```diff
@@ -0,0 +1,299 @@
+                        response.get("error_description"),
+                        response.get("error_uri"),
+                    ),
+                )
+            )
+            logger.error(error_message)
+            raise OAuth2TokenException(error_message)
+
+        data = {"access_token": access_token}
+
+        # The authorization server MAY issue a new refresh token, in which case
+        # the client MUST discard the old refresh token and replace it with the
+        # new refresh token.
+        #
+        # https://datatracker.ietf.org/doc/html/rfc6749#section-6
+
+        refresh_token = response.get("refresh_token")
+        if refresh_token:
+            data["refresh_token"] = refresh_token
+
+        # If omitted, the authorization server SHOULD provide the
+        # expiration time via other means or document the default value.
+        #
+        # https://datatracker.ietf.org/doc/html/rfc6749#section-5.1
+        #
+        # This alternate way of specifying the expiration is handled
+        # by the optional expires_in field of the OAuth2AuthenticationConfiguration
+
+        expires_in = response.get("expires_in") or self.expires_in
+        if expires_in:
+            data["expires_at"] = int(datetime.utcnow().timestamp()) + expires_in
+
+        # persist new tokens to the database
+        # ideally we use a passed in database session but we create a new
+        # session if one is not available
+        updated_secrets = {**connection_config.secrets, **data}
+        if db:
+            connection_config.update(db, data={"secrets": updated_secrets})
+        else:
+            SessionLocal = get_db_session()
+            new_db: Session = SessionLocal()
+            merged_connection_config = new_db.merge(connection_config)
+            merged_connection_config.update(new_db, data={"secrets": updated_secrets})
+            new_db.close()
+
+        logger.info(
+            f"Successfully updated the OAuth2 token(s) for {connection_config.key}"
+        )
+
+        return access_token
+
+    def get_authorization_url(
+        self, db: Session, connection_config: ConnectionConfig
+    ) -> Optional[str]:
+        """
+        Returns the authorization URL to initiate the OAuth2 workflow.
+        Also stores a reference between the authorization request and the connector
+        to be able to link the returned auth code to the correct connector."""
+
+        # make sure the required secrets to generate the authorization url have been provided
+        self._check_required_secrets(connection_config)
+
+        # generate the state that will be used to link this authorization request to this connector
+        state = str(uuid4())
+        AuthenticationRequest.create_or_update(
+            db, data={"connection_key": connection_config.key, "state": state}
+        )
+        # add state to secrets
+        connection_config.secrets["state"] = state
```

Comment:
The `state` is a little strange in that it has to be defined as a quer_param in the saas_config for /auth/authorize but we supply the state ourselves. If the user doesn't list that as a query param we won't return the state in the url we build.  Good documentation will help with this.

Reply:
I'll keep this in mind for the documentation ticket